### PR TITLE
feat-notification: Notification conversion step for harness

### PR DESCRIPTION
**Pull request by vanisrikanithi**
Notification go-conversion:
Added "Notification" support to go-convert specific to jenkins migration.

Norification Jenkins Trace:
[notification_28.json](https://github.com/user-attachments/files/18118924/notification_28.json)

go run main.go jenkinsjson --downgrade ~/go/notification_28.json

Harness Notification Pipeline is [here](https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/module/ci/orgs/default/projects/jenkinstoharness/pipelines/Vani_Notification/executions/bFVDXNsrQXmpeYWMwmdAuQ/pipeline).

Notification harness pipeline step:
```
- step:
    identifier: notifyendpointsaa7206
    name: Notification
    spec:
      image: plugins/webhook
      settings:
        content-type: application/json
        method: POST
        template: |-
          {
              "status": "SUCCESSFUL",
              "notes": "Build metrics for analysis"
          }
        token-value: <+input>
        urls:
          - https://webhook.site/9ffae84b-a338-43ef-9283-319d70574bf4
    timeout: ""
    type: Plugin
```

### CODE REVIEW

**Comment by vanisrikanithi** [comment]
Hello @Ompragash ,

Could you please review the Notification conversion PR and provide your comments if there are any?

Thank you.

**Review by Ompragash — CHANGES_REQUESTED** [review]
looks like the template value is being formatted with a newline, which will throw an error in Harness. To fix this, ensure the template value is on a single line in the generated YAML. Once fixed, the changes can be merged
```
              identifier: Stage_null87edb4
              name: Notify
              steps:
              - step:
                  identifier: notifyendpointsca9fc8
                  name: Notification
                  spec:
                    image: plugins/webhook
                    settings:
                      content_type: application/json
                      debug: "true"
                      method: <+input>
                      password: <+input>
                      template: '{"status": "Build Successful", "job": "${env.JOB_NAME}",
                        "buildNumber": "${env.BUILD_NUMBER}"}'
                      urls: <+input>
                      username: <+input>
                  timeout: ""
                  type: Plugin
```

**Comment by vanisrikanithi** [comment]
@Ompragash, YAML serializers typically have a default line length limit. If the string exceeds this limit, the serializer splits it into multiple lines.

Happy Path: I have copied the conversion step and pasted it in harness yaml, once I save the yaml it is automatically formatting the template.
[Harness Test Pipeline](https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/module/ci/orgs/default/projects/jenkinstoharness/pipelines/Vani_Notification/executions/8U7FlfM7SwqykqJkBxG1nA/pipeline)

Could you please check it and confirm the changes?

**Comment by vanisrikanithi** [comment]
@Ompragash, Hello OP, updated the Notify Endpoints as per our discussion. Could you please review it and let me know your comments on this?

File: convert/jenkinsjson/json/notification_test.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		Type: "plugin",
+		Spec: &harness.StepPlugin{
+			Image: "plugins/webhook",
+			With: map[string]interface{}{
+				"urls":         []string{"https://webhook.site/9ffae84b-a338-43ef-9283-319d70574bf4"},
+				"method":       "POST",
+				"username":     "<+input>",
+				"password":     "<+input>",
+				"token-value":  "<+input>",
+				"token-type":   "<+input>",
```
Comment (Ompragash) [inline convert/jenkinsjson/json/notification_test.go:25]: we don't need to mention both password and token at the same time - can you please update it to just token and remove username and password

**Review by Ompragash — CHANGES_REQUESTED** [review]

File: convert/jenkinsjson/json/notification.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	} else {
+		urls = append(urls, "<+input>")
+	}
+
+	phase, _ := parameterMap["phase"].(string)
+	notes, _ := parameterMap["notes"].(string)
+
+	// Create the template
+	template := fmt.Sprintf(`{
+  		"phase": "%s",
```
Comment (Ompragash) [inline convert/jenkinsjson/json/notification.go:27]:
```suggestion
  		"status": "%s",
```

whenever the phase is completed in jenkins trace, then the value of status should be successful - if it's not, then leave this empty

**Review by Ompragash — CHANGES_REQUESTED** [review]
Tests are failing - once it's fixed then we are good to merge
~~~
--- FAIL: TestConvertNotification (0.00s)
    base_test.go:27: failed to read JSON file: open /drone/src/convert/jenkinsjson/convertTestFiles/notification/notification_snippet.json: no such file or directory
Missing 'delegate' in parameterMap
No entries operations found in arguments
{"id":"slackSend8962b0","name":"slackSend","type":"plugin","spec":{"image":"plugins/slack","with":{"access_token":"slackcreds03","channel":"dev","color":"good","committers_slack_id":true,"icon_emoji":":rocket:","message":"hi, this is senthil-slack-file-send-03_dec10","username":"jenkins-test-user"},"inputs":{"access_token":"slackcreds03","channel":"dev","color":"good","committers_slack_id":true,"icon_emoji":":rocket:","message":"hi, this is senthil-slack-file-send-03_dec10","username":"jenkins-test-user"}}}
2024/12/17 00:34:35 withCredentials unsupported symbol: not_real
FAIL
coverage: 48.5% of statements
~~~

**Review by Ompragash — APPROVED** [review]

**Pull request merged**